presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometrySerde.java

```diff
@@ -44,58 +54,152 @@ private GeometrySerde() {}
-            while (input.available() > 0) {
-                geometries.add(readGeometry(isGeometryCollection, input));
-            }
+            int endPosition = output.size();
+            int length = endPosition - startPosition - Integer.BYTES;
+
+            output.getUnderlyingSlice().setInt(startPosition, length);
         }
+    }
+
+    private static void writeSimpleGeometry(DynamicSliceOutput output, GeometryType type, OGCGeometry geometry)
+    {
+        output.appendByte(type.code());
+        Geometry esriGeometry = requireNonNull(geometry.getEsriGeometry(), "esriGeometry is null");
+        byte[] shape = geometryToEsriShape(esriGeometry);
+        output.appendBytes(shape);
+    }
 
-        if (geometries.isEmpty()) {
-            return new OGCConcreteGeometryCollection(emptyList(), null);
+    public static OGCGeometry deserialize(Slice shape)
+    {
+        requireNonNull(shape, "shape is null");
+        BasicSliceInput input = shape.getInput();
+        verify(input.available() > 0);
+        int length = input.available() - 1;
+        GeometryType type = GeometryType.getForCode(input.readByte());
+        return readGeometry(input, shape, type, length);
+    }
+
+    private static OGCGeometry readGeometry(BasicSliceInput input, Slice inputSlice, GeometryType type, int length)
+    {
+        switch (type) {
+            case POINT:
+            case MULTI_POINT:
+            case LINE_STRING:
+            case MULTI_LINE_STRING:
+            case POLYGON:
+            case MULTI_POLYGON:
+                return readSimpleGeometry(input, inputSlice, type, length);
+            case GEOMETRY_COLLECTION:
+                return readGeometryCollection(input, inputSlice);
+            default:
+                throw new UnsupportedOperationException("Unexpected type: " + type);
         }
-        else if (geometries.size() == 1) {
-            return geometries.get(0);
+    }
+
+    private static OGCConcreteGeometryCollection readGeometryCollection(BasicSliceInput input, Slice inputSlice)
+    {
+        // GeometryCollection: geometryType|len-of-shape1|bytes-of-shape1|len-of-shape2|bytes-of-shape2...
+        List<OGCGeometry> geometries = new ArrayList<>();
+        while (input.available() > 0) {
+            int length = input.readInt() - 1;
+            GeometryType type = GeometryType.getForCode(input.readByte());
+            geometries.add(readGeometry(input, inputSlice, type, length));
         }
         return new OGCConcreteGeometryCollection(geometries, null);
     }
 
-    private static OGCGeometry readGeometry(boolean isGeometryCollection, BasicSliceInput input)
+    private static OGCGeometry readSimpleGeometry(BasicSliceInput input, Slice inputSlice, GeometryType type, int length)
+    {
+        int currentPosition = toIntExact(input.position());
+        ByteBuffer geometryBuffer = inputSlice.toByteBuffer(currentPosition, length).slice();
```

Comment:
Why is the slice() call needed here?

Reply:
`position` is ignored. Underlying algorithm uses `put*(index, value)` methods instead of `put*(value)`. So a new byte buffer with `offset` set is needed